iib/web/api_v1.py

```diff
@@ -460,3 +460,108 @@ def regenerate_bundle_batch():
         ', '.join(request_id_strs),
     )
     return flask.jsonify(request_jsons), 201
+
+
+@api_v1.route('/builds/add-rm-batch', methods=['POST'])
+@login_required
+def add_rm_batch():
+    """
+    Submit a batch of requests to add or remove operators from an index image.
+
+    :rtype: flask.Response
+    :raise ValidationError: if required parameters are not supplied
+    """
+    payload = flask.request.get_json()
+    Batch.validate_batch_request_params(payload)
+
+    batch = Batch(annotations=payload.get('annotations'))
+    db.session.add(batch)
+
+    requests = []
+    # Iterate through all the build requests and verify that the requests are valid before
+    # committing them and scheduling the tasks
+    for build_request in payload['build_requests']:
+        try:
+            if build_request.get('operators'):
+                # Check for the validity of a RM request
+                request = RequestRm.from_json(build_request, batch)
+            elif build_request.get('bundles'):
+                # Check for the validity of an Add request
+                request = RequestAdd.from_json(build_request, batch)
+            else:
+                raise ValidationError('Build request is not a valid Add/Rm request.')
+        except ValidationError as e:
+            # Rollback the transaction if any of the build requests are invalid
+            db.session.rollback()
```

Comment:
I'm curious if this is actually needed given that you're just about to exit the request handler anyway?  Is there something in place to automatically rollback if a request handler raises (or if not rollback, just drop the session without committing)?

If it is needed to explicitly rollback whenever a request is about to fail: that seems awfully fragile.  It's not like ValidationError is the only kind of error which could possibly occur.  Why is it appropriate to roll back on that exception and not others?

Based on this text in https://flask-sqlalchemy.palletsprojects.com/en/2.x/quickstart/ I'd have thought that this code isn't useful:

> You have to commit the session, but you don’t have to remove it at the end of the request, Flask-SQLAlchemy does that for you.

Reply:
Nice catch! I believe I can remove that piece of code.

/cc @mprahl @lcarva

Reply:
So Flask-SQLAlchemy will call `session.remove` when the Flask application context is torn down:
https://github.com/pallets/flask-sqlalchemy/blob/7352d8b941b397b58341667c455c989baba16cd1/src/flask_sqlalchemy/__init__.py#L852

This ends up calling `session.close` which rolls back the transaction as described here:
https://docs.sqlalchemy.org/en/13/orm/contextual.html#using-thread-local-scope-with-web-applications

So yes, the line was unnecessary but it was also harmless.